README.md
# multiviewtest
R package that implements tests for whether clusterings of two data views (both multivariate, both networks, or one multivariate and one network) are independent. See reference 1: Gao, L.L., Bien, J., and Witten, D. (2020) "Are Clusterings of Multiple Data Views Independent?", Biostatistics, DOI: 10.1093/biostatistics/kxz001 and reference 2:  Gao, L.L., Witten, D., Bien, J. (2021) Testing for Association in Multi-View Network Data, to appear in Biometrics, DOI: 10.1111/biom.13464,

NEWS.md
# multiviewtest 2.0.1
1. Changed a few lines to make the package compatible with the upcoming version of mclust. 



R/test_indep_com.R
#' Pseudo pseudolikelihood ratio test for dependent communities
#'
#'
#' Implements the pseudo pseudolikelihood ratio test described in Section 3 of
#' Gao et. al. (2019) "Testing for Association in Multi-View Network Data"
#' for testing for dependence between communities in two network data views.
#' Fits stochastic block models in each view.
#'
#' @param X Multi-view data with two views; a list of two n x n adjacency matrices.
#' @param nperm An integer specifying the number of permutations to
#' use for the permutation procedure. The default number is 200.
#' @param step A numeric value containing the fixed step size to be used in the optimization
#' algorithm for estimating Pi. The default step size is 0.001. 
#' @param maxiter A numeric value containing the maximum number of iterations to run in
#'  the optimization algorithm. The default maximum is 1000.
#' @param K1 An optional argument containing the number of communities in View 1.
#'  If left out, then the number of communities is chosen with the method of 
#'  Le and Levina (2015).
#' @param K2 An optional argument containing the number of communities in View 2.
#'  If left out, then the number of communities is chosen with the method of 
#'  Le and Levina (2015). 
#' @param parallel An optional argument; if true, do parallel computing using the 
#' doParallel package 
#' @import randnet
#' @importFrom foreach %dopar%
#' @importFrom foreach foreach
#' @export
#' @return
#' A list containing the following output components:
#' \item{K1}{The number of communities in view 1}
#' \item{K2}{The number of communities in view 2}
#' \item{Pi.est}{The estimated Pi matrix}
#' \item{P2LRstat}{The pseudo likelihood ratio test statistic}
#' \item{pval}{The p-value}
#' \item{modelfit1}{The parameter estimates and community assignment estimates from View 1.}
#' \item{modelfit2}{The parameter estimates and community assignment estimates from View 2.}
#' @examples
#' set.seed(1)
#' n <- 50
#' Pi <- diag(c(0.5, 0.5))
#' theta1 <- rbind(c(0.5, 0.1), c(0.1, 0.5))
#' theta2 <- cbind(c(0.1, 0.5), c(0.5, 0.1))
#' 
#' # 50 draws from a multi-view SBM with perfectly dependent communities
#' dat <- mv_sbm_gen(n, Pi, theta1, theta2)
#' 
#' # Test H0: communities are independent
#' # Data was generated under the alternative hypothesis 
#' results <- test_indep_com(dat$data, nperm=25)
#' results$pval
#'
#' @references
#' Amini, A. A., Chen, A., Bickel, P. J., & Levina, E. (2013). Pseudo-likelihood methods 
#' for community detection in large sparse networks. The Annals of Statistics, 41(4), 2097-2122.
#'
#' Gao, L.L., Witten, D., Bien, J. Testing for Association in Multi-View Network Data, preprint. 
#'
#' Le, C. M., & Levina, E. (2015). Estimating the number of communities 
#' in networks by spectral methods. arXiv preprint arXiv:1507.00827.
test_indep_com <- function(X, K1=NULL, K2=NULL, nperm=200, step=0.001, maxiter=1000, 
                           parallel=FALSE) {
  input_check <- function(X, K1, K2, nperm, step, maxiter, parallel) { 
    if(typeof(X) != "list") {
      stop("X is not in the right format; should be a list of two matrices")
    } 
    
    if(length(X) != 2) {
      stop("X is not in the right format; should be a list of two matrices")
    }
   
    if(!is.matrix(X[[1]]) | !is.numeric(X[[1]]) | !is.matrix(X[[2]]) | !is.numeric(X[[2]])) {
      stop("X is not in the right format; should be a list of two matrices")
    } 
    
    if(nrow(X[[1]]) != ncol(X[[1]]) | nrow(X[[2]]) != ncol(X[[2]]) | nrow(X[[1]]) != ncol(X[[2]])) { 
      stop("X is not in the right format; should be a list of two n x n matrices")  
    } 
    
    if(!isSymmetric(X[[1]]) | !isSymmetric(X[[2]])) { 
      stop("Networks should be undirected, so adjacency matrices should be symmetric")        
    }
    
    if(any(is.na(X[[1]])) || any(is.na(X[[2]])))
      stop(paste("Views cannot contain any missing values"))
    
    if(step <= 0) stop('Step size for optimization algorithm should be > 0')
    if(nperm < 1) stop('Number of permutations should be > 0')
    if(nperm <= 50) warning('Number of permutation iterations is specified to be small;
                     p-value approximation will likely be imprecise')
    if(maxiter %% 1 != 0|| maxiter <= 0) stop('Maximum number of iterations should be
                                                a positive integer')
    
    if(!is.null(K1)) { 
      if(!is.numeric(K1)) stop('Number of communities must be a positive integer in View 1')
      if(K1 %% 1 != 0 || K1 < 0) stop('Number of communities must be a positive integer in View 1')
      if(K1 < 1) stop('Number of communities must be greater than 1 in View 1')
    }
    
    if(!is.null(K2)) { 
      if(!is.numeric(K2)) stop('Number of communities must be a positive integer in View 2')
      if(K2 %% 1 != 0 || K2 < 0) stop('Number of communities must be a positive integer in View 2')
      if(K2 < 1) stop('Number of communities must be greater than 1 in View 2')
    }
    
    if(!is.logical(parallel)) stop("parallel should be TRUE or FALSE")
  } 
  
  input_check(X, K1, K2, nperm, step, maxiter, parallel)
  
  if(is.null(K1)) K1 <- max(randnet::BHMC.estimate(X[[1]])$K, 2)  
  if(is.null(K2)) K2 <- max(randnet::BHMC.estimate(X[[2]])$K, 2)  
  
  cluster.results1 <- pl_est_com(X[[1]], K1, parallel=parallel)
  cluster.results2 <- pl_est_com(X[[2]], K2, parallel=parallel)
  
  logphi1 <- cluster.results1$logphi
  logphi2 <- cluster.results2$logphi
  pi1.h <- cluster.results1$pi
  pi2.h <- cluster.results2$pi 

  # Computes diag(pi1.h)*C.h*diag(pi2.h)
  Pi.hat <- optimize_over_Pi(logphi1, logphi2, pi1.h, pi2.h,
                                 max.iter=maxiter, stepsz=step)

  # Compute P2LRT
  null.pll <- cluster.results1$ploglik + cluster.results2$ploglik 
  p2lr.stat <- (-Pi.hat$obj) - null.pll
  n <- nrow(X[[1]])
  
  do_one_perm <- function() { 
    perm <- sample(1:n, replace=F)
    Pi.hat.perm <- optimize_over_Pi(logphi1, logphi2[perm, ], 
                                     pi1.h, pi2.h, max.iter=maxiter, stepsz=step)
    -Pi.hat.perm$obj - null.pll
  }
  
  if(nperm <= 1000) { 
    if(parallel) { 
      perms <- foreach(i = 1:nperm, .combine=c) %dopar% do_one_perm()
    } else { 
      perms <- replicate(nperm, do_one_perm())  
      perms <- perms[!is.na(perms)] # remove any NA's
      pval <- (sum(perms >= p2lr.stat) + 1)/(length(perms) + 1)
      return(list(K1=K1, K2=K2, Pi.est=Pi.hat$Pi, P2LRstat=p2lr.stat, 
                  pval=pval, modelfit1=cluster.results1, modelfit2=cluster.results2))
    }
  } else { 
    if(parallel) { 
      interim.perm <- foreach::foreach(i = 1:1000, .combine=c) %dopar% do_one_perm()
    } else { 
      interim.perm <- replicate(1000, do_one_perm())  
    }
    
    interim.perm <- interim.perm[!is.na(interim.perm)] # Remove any NA's 
    interim.pval <- (sum(interim.perm >= p2lr.stat) + 1)/(length(interim.perm) + 1) 
  
    if(interim.pval > 0.1) { 
      cat("Stopped at 1000 permutations (p-value > 0.1)")
      return(list(K1=K1, K2=K2, Pi.est=Pi.hat$Pi, P2LRstat=p2lr.stat, 
                  pval=interim.pval, modelfit1=cluster.results1, modelfit2=cluster.results2))
    } 
  
    if(parallel) { 
      rest.perm <- foreach::foreach(i = 1001:nperm, .combine=c) %dopar% do_one_perm()
    } else { 
      rest.perm <- replicate(nperm - 1000 + 1, do_one_perm())  
    }
    
    final.perm <- c(interim.perm, rest.perm)
    final.perm <- final.perm[!is.na(final.perm)] # Remove any NA's 
    pval <- (sum(final.perm >= p2lr.stat) + 1)/(length(final.perm) + 1)
    
    return(list(K1=K1, K2=K2, Pi.est=Pi.hat$Pi, P2LRstat=p2lr.stat, 
                pval=pval, modelfit1=cluster.results1, modelfit2=cluster.results2))
  }
}

R/test_indep_com_clust.R
#' Pseudo pseudolikelihood ratio test for association between communities in 
#' a network view and clusters in a multivariate view
#'
#'
#' Implements the pseudo pseudolikelihood ratio test described in Section 4 of
#' Gao et. al. (2019) "Testing for Association in Multi-View Network Data"
#' for testing for dependence between communities in a network data view
#' and cluster in a multivariate view. Fits a stochastic block model in the network view, 
#' and a Gaussian mixture model in the multivariate view. 
#'
#' @param X Multi-view data with two views; a list of two n x n adjacency matrices.
#' @param nperm An integer specifying the number of permutations to
#' use for the permutation procedure. The default number is 200.
#' @param step A numeric value containing the fixed step size to be used in the optimization
#' algorithm for estimating Pi. The default step size is 0.001. 
#' @param maxiter A numeric value containing the maximum number of iterations to run in
#'  the optimization algorithm. The default maximum is 1000.
#' @param K1 An optional argument containing the number of communities in View 1.
#'  If left out, then the number of communities is chosen with the method of 
#'  Le and Levina (2015).
#' @param K2 An optional argument containing the number of clusters in View 2.
#'  If left out, then the number of clusters is chosen with BIC. 
#' @param model2 A character string indicating the model
#' to be fitted for Gaussian model-based clustering in the multivariate view using
#' the function \code{\link[mclust]{Mclust}}. The default is \code{"EII"}
#' (spherical, equal volume).
#' The help file for \code{\link[mclust]{mclustModelNames}} describes
#' the available model options.
#' @param init2 An optional argument containing the model to be fitted in the
#' hierarchical clustering initialization in Gaussian model-based clustering
#' in in the multivariate view . The default is \code{"VVV"} (ellipsoidal, 
#' varying volume, shape, and orientation). The help file for \code{\link[mclust]{hc}} 
#' describes the available model options.
#' @param parallel An optional argument allowing for parallel computing using the 
#' doParallel package 
#' @import mclust
#' @import randnet
#' @importFrom foreach %dopar%
#' @importFrom foreach foreach
#' @export
#' @return
#' A list containing the following output components:
#' \item{K1}{The number of communities in view 1}
#' \item{K2}{The number of communities in view 2}
#' \item{Pi.est}{The estimated Pi matrix}
#' \item{P2LRstat}{The pseudo likelihood ratio test statistic}
#' \item{pval}{The p-value}
#' \item{modelfit1}{The parameter estimates and community assignment estimates from View 1.}
#' \item{modelfit2}{The parameter estimates and community assignment estimates from View 2.}
#' @examples
#' # 50 draws from a multi-view SBM, where the clusters 
#' # and the communities are independent
#' n <- 50
#' Pi <- tcrossprod(c(0.5, 0.5), c(0.5, 0.5))
#' theta1 <- rbind(c(0.5, 0.1), c(0.1, 0.5))
#' mu2 <- cbind(c(2, 2), c(-2, 2))
#' Sigma2 <- diag(rep(0.5, 2))
#' 
#' dat <- mv_sbm_gmm_gen(n, Pi, theta1, mu2, Sigma2)
#' 
#' 
#' # Test H0: communities are independent
#' # Data was generated under the null hypothesis
#' results <- test_indep_com_clust(dat$data, nperm=25)
#' results$pval
#'
#' @references
#' Amini, A. A., Chen, A., Bickel, P. J., & Levina, E. (2013). Pseudo-likelihood methods 
#' for community detection in large sparse networks. The Annals of Statistics, 41(4), 2097-2122.
#' 
#' Fraley C. and Raftery A. E. (2002) Model-based clustering,
#' discriminant analysis and density estimation,
#' Journal of the American Statistical Association, 97/458, pp. 611-631. 
#
#' Gao, L.L., Witten, D., Bien, J. Testing for Association in Multi-View Network Data, preprint. 
#' 
#' Le, C. M., & Levina, E. (2015). Estimating the number of communities 
#' in networks by spectral methods. arXiv preprint arXiv:1507.00827.
#' 
#' Scrucca L., Fop M., Murphy T. B. and Raftery A. E. (2016)
#' mclust 5: clustering, classification and density estimation
#' using Gaussian finite mixture models, The R Journal, 8/1, pp. 205-233.
test_indep_com_clust <- function(X, K1=NULL, K2=NULL, model2="EII", 
                                 init2=NULL, nperm=200, step=0.001, 
                                 maxiter=1000, parallel=FALSE) {
  input_check <- function(X, K1, K2, model2, init2, nperm, step, 
                          maxiter, parallel) { 
    if(typeof(X) != "list") {
      stop("X is not in the right format; should be a list of two matrices")
    } 
    
    if(length(X) != 2) {
      stop("X is not in the right format; should be a list of two matrices")
    }
    
    if(!is.matrix(X[[1]]) | !is.numeric(X[[1]]) | !is.matrix(X[[2]]) | !is.numeric(X[[2]])) {
      stop("X is not in the right format; should be a list of two matrices")
    } 
    
    if(nrow(X[[1]]) != ncol(X[[1]])) { 
      stop("X is not in the right format; adjacency matrix should be n x n matrix")  
    } 
    
    if(!isSymmetric(X[[1]])) { 
      stop("Network should be undirected, so adjacency matrix should be symmetric")        
    }
    
    if(any(is.na(X[[1]])) || any(is.na(X[[2]])))
      stop(paste("Views cannot contain any missing values"))
    
    if(!model2 %in% c("E", "V", "EII", "VII", "EEI", "VEI", 
                      "EVI", "VVI", "EEE", "EVE", "VEE", "VVE", 
                      "EEV", "VEV", "EVV", "VVV"))
      stop(paste("Did not pass a valid model for Mclust in multivariate view") )
    
    if(model2 %in% c("E", "V") && is.matrix(X[[2]]))
      stop(paste(model2, "is a univariate model for Mclust, cannot be used for
               multivariate data"))
    if(model2 %in% c("E", "V", "EII", "VII", "EEI", "VEI", 
                     "EVI", "VVI", "EEE", "EVE", "VEE", "VVE", 
                     "EEV", "VEV", "EVV", "VVV") && !is.matrix(X[[2]]))
      stop(paste(model2, "is a multivariate model for Mclust, cannot be used for
         univariate data") )
    
    if(!is.null(init2)) {
      if(!init2 %in% c("E", "V", "EII", "VII", "EEE", "VVV"))
        stop(paste("Did not pass a valid initialization model for Mclust in multivariate view") )
      
      if(init2 %in% c("E", "V") && is.matrix(X[[2]]))
        stop(paste(init2, "is a univariate initialization model for Mclust,
                   cannot be used for multivariate data"))
      if(init2 %in% c("EII", "VII", "EEE", "VVV") && !is.matrix(X[[2]]))
        stop(paste(init2, "is a univariate initialization model for Mclust,
                   cannot be used for univariate data") )
      agree2 <- ((init2 %in% c("E", "V")) == (model2 %in% c("E", "V")))
      if(!agree2) {
        stop("In multivariate view, initialization model is for univariate data,
             but model is for multivariate data, or
             initialization model is for multivariate data,
             but model is for univariate data")
      }
    }
    
    if(step <= 0) stop('Step size for optimization algorithm should be > 0')
    if(nperm < 1) stop('Number of permutations should be > 0')
    if(nperm <= 50) warning('Number of permutation iterations is specified to be small;
                            p-value approximation will likely be imprecise')
    if(maxiter %% 1 != 0|| maxiter <= 0) stop('Maximum number of iterations should be
                                              a positive integer')
    
    if(!is.null(K1)) { 
      if(!is.numeric(K1)) stop('Number of communities must be a positive integer')
      if(K1 %% 1 != 0 || K1 < 0) stop('Number of communities must be a positive integer')
      if(K1 < 1) stop('Number of communities must be greater than 1')
    }
    
    if(!is.null(K2)) { 
      if(!is.numeric(K2)) stop('Number of clusters must be a positive integer')
      if(K2 %% 1 != 0 || K2 < 0) stop('Number of clusters must be a positive integer')
      if(K2 < 1) stop('Number of clusters must be greater than 1')
    }    
    
    if(!is.logical(parallel)) stop("parallel should be TRUE or FALSE")
  } 
  
  input_check(X, K1, K2, model2, init2, nperm, step, maxiter, parallel)
  
  if(is.null(K1)) K1 <- max(randnet::BHMC.estimate(X[[1]])$K, 2)  
  cluster.results1 <- pl_est_com(X[[1]], K1, parallel=parallel)
  logphi1 <- cluster.results1$logphi
  pi1.h <- cluster.results1$pi
  n <- nrow(logphi1)
  
  if(is.null(init2) & is.matrix(X[[2]]))  init2 <- "VVV"
  if(is.null(init2) & !is.matrix(X[[2]]))  init2 <- "V"
  
  if(is.null(K2)) {
    EM.View2 <- mclust::Mclust(X[[2]], G=2:9, modelNames=c(model2),
                               initialization=list(hcPairs=hc(X[[2]],
                                                              modelName=init2)))
    if(is.null(EM.View2)) stop("The model specified for model-based clustering
                               could not be fitted in view 2")
    K2 <- EM.View2$G
  } else {
    EM.View2 <- mclust::Mclust(X[[2]], K2, modelNames=c(model2),
                               initialization=list(hcPairs=hc(X[[2]],
                                                              modelName=init2)))
    if(is.null(EM.View2)) stop("The model specified for model-based clustering
                               could not be fitted in view 2")
  }
  
  EM.View2.param <- EM.View2$parameters


  logphi2 <- mclust::cdens(modelName=model2, data=X[[2]], logarithm=TRUE, parameters=EM.View2.param)
  pi2.h <- EM.View2.param$pro
  
  # Computes diag(pi1.h)*C.h*diag(pi2.h)
  Pi.hat <- optimize_over_Pi(logphi1, logphi2, pi1.h, pi2.h,
                             max.iter=maxiter, stepsz=step)
  
  # Compute P2LRT
  null.pll <- cluster.results1$ploglik + EM.View2$loglik
  p2lr.stat <- (-Pi.hat$obj) - null.pll
  
  do_one_perm <- function() { 
    perm <- sample(1:n, replace=F)
    Pi.hat.perm <- optimize_over_Pi(logphi1, logphi2[perm, ], 
                                    pi1.h, pi2.h, max.iter=maxiter, stepsz=step)
    -Pi.hat.perm$obj - null.pll
  }
  
  if(nperm <= 1000) { 
    if(parallel) { 
      perms <- foreach(i = 1:nperm, .combine=c) %dopar% do_one_perm()
    } else { 
      perms <- replicate(nperm, do_one_perm())  
      perms <- perms[!is.na(perms)] # remove any NA's
      pval <- (sum(perms >= p2lr.stat) + 1)/(length(perms) + 1)
      return(list(K1=K1, K2=K2, Pi.est=Pi.hat$Pi, P2LRstat=p2lr.stat, 
                  pval=pval, modelfit1=cluster.results1, modelfit2=EM.View2))
    }
  } else { 
    if(parallel) { 
      interim.perm <- foreach::foreach(i = 1:1000, .combine=c) %dopar% do_one_perm()
    } else { 
      interim.perm <- replicate(1000, do_one_perm())  
    }
    
    interim.perm <- interim.perm[!is.na(interim.perm)] # Remove any NA's 
    interim.pval <- (sum(interim.perm >= p2lr.stat) + 1)/(length(interim.perm) + 1) 
    
    if(interim.pval > 0.1) { 
      cat("Stopped at 1000 permutations (p-value > 0.1)")
      return(list(K1=K1, K2=K2, Pi.est=Pi.hat$Pi, P2LRstat=p2lr.stat, 
                  pval=interim.pval, modelfit1=cluster.results1, modelfit2=EM.View2))
    } 
    
    if(parallel) { 
      rest.perm <- foreach::foreach(i = 1001:nperm, .combine=c) %dopar% do_one_perm()
    } else { 
      rest.perm <- replicate(nperm - 1000 + 1, do_one_perm())  
    }
    
    final.perm <- c(interim.perm, rest.perm)
    final.perm <- final.perm[!is.na(final.perm)] # Remove any NA's 
    pval <- (sum(final.perm >= p2lr.stat) + 1)/(length(final.perm) + 1)
    
    return(list(K1=K1, K2=K2, Pi.est=Pi.hat$Pi, P2LRstat=p2lr.stat, 
                pval=pval, modelfit1=cluster.results1, modelfit2=EM.View2))
  }
}

R/mv_gmm_gen.R
#' Generates data from a multi-view Gaussian mixture model
#'
#' Generates data from a multi-view Gaussian mixture model with n observations 
#' and two views. 
#'
#' @param n number of observations
#' @param Pi K1 x K2 matrix where the (k, k')th entry contains the probability of an observation 
#' belonging to cluster k in View 1 and cluster k' in View 2
#' @param mu1 p1 x K1 matrix where the columns contain the K1 cluster means in View 1
#' @param mu2 p2 x K2 matrix where the columns contain the K2 cluster means in View 2
#' @param Sigma1 p1 x p1 matrix containing the covariance matrix for View 1
#' @param Sigma2 p2 x p2 matrix containing the covariance matrix for View 2
#'
#' @importFrom stats rnorm
#' @export
#' @return 
#' A list containing the following components:
#' \item{data}{A list with two items: the view 1 n x p1 multivariate data set and 
#' the view 2 n x p2 multivariate data set }
#' \item{clusters}{A list with two items: the view 1 cluster memberships and 
#' the view 2 cluster memberships}
#' 
#' @examples
#' # 25 draws from a two-view Gaussian mixture model where the clusters are independent
#' n <- 25
#' Pi <-  tcrossprod(c(0.5, 0.5), c(0.25, 0.25, 0.5))
#' mu1 <- cbind(c(2, 2), c(-2, 2))
#' mu2 <- cbind(c(0, 1), c(1, 0), c(-1, 0))
#' Sigma1 <- diag(rep(1, 2))
#' Sigma2 <- diag(rep(0.5, 2))
#' 
#' mv_gmm_gen(n, Pi, mu1, mu2, Sigma1, Sigma2)
#' 
#' @references
#' Gao, L.L., Bien, J., Witten, D. (2019) Are Clusterings of Multiple Data Views Independent?
#' Biostatistics,  <DOI:10.1093/biostatistics/kxz001>
#'
#' Gao, L.L., Witten, D., Bien, J. Testing for Association in Multi-View Network Data, preprint. 
mv_gmm_gen <- function(n, Pi, mu1, mu2, Sigma1, Sigma2) {
  input_check <- function(n, Pi, mu1, mu2, Sigma1, Sigma2) { 
    if(!is.matrix(Pi) | !is.numeric(Pi)) {
      stop("Pi must be a numeric matrix")  
    } 
    
    if(any(Pi < 0) | (abs(sum(Pi) - 1) > 1e-15)) { 
        stop("Pi must be non-negative and sum to 1")  
    }  
    
    if(!is.numeric(n)) { 
      stop("n must be a positive integer")
    } 
    
    if((n %% 1 != 0) | n <= 0) { 
        stop("n must be a positive integer")  
    }
    
    if(!is.matrix(mu1) | !is.numeric(mu1) | !is.matrix(mu2) | !is.numeric(mu2)) { 
      stop("mu1 and mu2 must be numeric matrices")
    } else { 
      if((ncol(mu1) != nrow(Pi)) | (ncol(mu2) != ncol(Pi))) { 
        stop("mu1 must be n x K1 and mu2 must be n x K2, where Pi is K1 x K2")  
      }  
    }
    
    if(!is.matrix(Sigma1) | !is.numeric(Sigma1) | !is.matrix(Sigma2) | !is.numeric(Sigma2)) { 
      stop("Sigma1 and Sigma2 must be numeric matrices")  
    } 
    
    if((nrow(Sigma1) != ncol(Sigma1)) | (nrow(Sigma2) != ncol(Sigma2))) { 
        stop("Sigma1 and Sigma2 must be PSD")  
    } 
    
    if(any(eigen(Sigma1)$values < 0) | any(eigen(Sigma2)$values < 0)) { 
          stop("Sigma1 and Sigma2 must be PSD")  
    }
  }
  
  generate_sv_gaussian <- function(mu, Sig, cl) {
    n <- length(cl)
    p <- nrow(Sig)
    
    matrix(stats::rnorm(n * p), n, p)%*%chol(Sig) + t(mu[, cl])
  }
  
  input_check(n, Pi, mu1, mu2, Sigma1, Sigma2)
  
  cl <- mv_memberships_gen(n, Pi)
  X1 <- generate_sv_gaussian(mu1, Sigma1, cl[, 1])
  X2 <- generate_sv_gaussian(mu2, Sigma2, cl[, 2])
  
  return(list(data=list(view1=X1, view2=X2), clusters=list(view1=cl[, 1], view2=cl[, 2])))
}



R/pl_est_com.R
#' Fits the stochastic block model using maximum pseudolikelihood estimation
#'
#' Fits the stochastic block model using maximum
#' pseudolikelihood estimation, as proposed by Amini et. al. (2013). 
#' This function implements the conditional pseudolikelihood algorithm 
#' from Amini et al. (2013). 
#'
#' @param X n x n adjacency matrix
#' @param K number of communities; by default, chosen using the method of Le and Levina (2015)
#' @param max.iter maximum number of iterations for the EM algorithm 
#' @param tol the EM algorithm stops when the relative tolerance is less than this value
#' @param parallel An optional argument allowing for parallel computing using the 
#' doParallel package 
#'
#' @import randnet
#' @import irlba
#' @import doParallel
#' @import matrixStats
#' @importFrom foreach %dopar%
#' @importFrom foreach foreach
#' @importFrom stats kmeans
#' @export
#' @return 
#' A list containing the following components:
#' \item{eta}{Estimate of eta, a K x K matrix defined in Amini et. al. (2013)}
#' \item{pi}{Estimate of the community membership probabilities}
#' \item{ploglik}{The maximum of the pseudolikelihood function}
#' \item{logphi}{n x K matrix, where (i, k)th entry contains the log p.m.f. of a multinomial
#' random variable with probability vector eta_k (the kth row of eta), evaluated at b_i, 
#' which is the ith row of the block compression matrix defined in Amini et. al. (2013)}
#' \item{responsibilities}{n x K matrix containing the responsibilities/soft cluster 
#' memberships for the n nodes}
#' \item{class}{A vector containing n (hard) cluster memberships for the n nodes}
#' \item{converged}{whether the algorithm converged to the desired tolerance}
#' 
#' @examples
#' # 50 draws from a stochastic block model for two network data views 
#' # where the communities are dependent
#' n <- 50
#' Pi <- diag(c(0.5, 0.5))
#' theta1 <- rbind(c(0.5, 0.1), c(0.1, 0.5))
#' theta2 <- cbind(c(0.1, 0.5), c(0.5, 0.1))
#' 
#' dat <- mv_sbm_gen(n, Pi, theta1, theta2)
#' 
#' # Fit SBM to view 1
#' results <- pl_est_com(X=dat$data$view1, K = 2)
#' table(results$class, dat$communities$view1)
#' 
#' @references
#' Amini, A. A., Chen, A., Bickel, P. J., & Levina, E. (2013). 
#' Pseudo-likelihood methods for community detection in large sparse networks. 
#' The Annals of Statistics, 41(4), 2097-2122.
#' 
#' Le, C. M., & Levina, E. (2015). Estimating the number of communities 
#' in networks by spectral methods. arXiv preprint arXiv:1507.00827.

pl_est_com <- function(X, K=NULL, max.iter = 1000, tol = 1e-8, parallel=FALSE) {
  input_check <- function(X, K, max.iter, tol, parallel) { 
    if(!is.numeric(X) | !is.numeric(max.iter) | !is.numeric(tol)) { 
      stop("All inputs must be numeric")  
    }
    
    if(length(max.iter) > 1 | length(tol) > 1) { 
      stop("max.iter and tol should be scalars")  
    }
    
    if(is.vector(X)) stop("X is not in the right format; should be a n x n matrix") 
    if(nrow(X) != ncol(X)) stop("X is not in the right format; should be a n x n matrix") 
    if(!isSymmetric(X)) stop("X must be a symmetric n x n matrix")
    if(!is.null(K)) { 
      if(!is.numeric(K)) stop("Number of communities must be a positive integer")
      if(length(K) > 1) stop("K should be scalar") 
      if(K %% 1 != 0 | K < 0) stop('Number of communities must be a positive integer')
      if(K < 1) stop('Number of communities must be greater than 1')
    }
    if(max.iter %% 1 != 0 | max.iter <= 0) stop("max number of iterations must be a positive integer")
    if(tol <= 0) stop("tolerance for convergence criteria must be positive")
    if(!is.logical(parallel)) stop("parallel should be T or F")
  }
  
  input_check(X, K, max.iter, tol, parallel)
  
  # if K is null, estimate using spectral method
  if(is.null(K)) K <- max(randnet::BHMC.estimate(X)$K, 2)
  
  # Zero degree nodes
  zd <- rowSums(X) == 0
  if(K > sum(!zd)) stop("Number of communities should be greater than the number of non-zero nodes.")
  
  # Save original data 
  orig.X <- X 
  orig.n <- nrow(X)
  
  # Work on data with zero degree nodes removed - we will handle the 
  # zero degree nodes separately at the end
  X <- X[!zd, !zd]
  n <- nrow(X) 
  
  # A parallelizable implementation of spectral clustering with perturbations
  reg.SP.parallel <- function(A, K, tau=0.25, lap=TRUE, parallel=TRUE) { 
    avg.d <- mean(colSums(A))
    A.tau <- A + tau * avg.d/nrow(A)
    if (!lap) {
      SVD <- irlba::irlba(A.tau, nu = K, nv = K)
    } else {
      d.tau <- colSums(A.tau)
      L.tau <- diag(1/sqrt(d.tau)) %*% A.tau %*% diag(1/sqrt(d.tau))
      SVD <- irlba::irlba(L.tau, nu = K, nv = K)
    }
    
    one_kmeans <- function() { 
      km <- kmeans(SVD$v[, 1:K], centers=K, nstart=1, iter.max=30)  
      return(list(cluster=km$cluster, loss=km$tot.withinss))
    }
    
    if(parallel) { 
      comb <- function(x, ...) {
        lapply(seq_along(x),
               function(i) c(x[[i]], lapply(list(...), function(y) y[[i]])))
      }
      
      km.res <- foreach::foreach(i = 1:30, .combine='comb', .multicombine=TRUE,
                      .init=list(list(), list())) %dopar% one_kmeans()
      best.km.res <- list(cluster=unlist(km.res[[1]][which.min(unlist(km.res[[2]]))]), 
                          loss=min(unlist(km.res[[2]])))
    } else { 
      km.res <- replicate(30, one_kmeans())
      best.km.res <- list(cluster=km.res[1, which.min(km.res[2, ])]$cluster, 
                          loss=km.res[2, which.min(km.res[2, ])]$loss)
    }
    
    return(list(cluster=best.km.res$cluster, loss=best.km.res$loss))
  }
  
  # Initialize the pseudolikelihood algorithm
  if(K < n) { 
    Z.hat <- reg.SP.parallel(X, K, parallel=parallel)$cluster 
  } else { 
    Z.hat <- seq(1, n)  
  } 
  
  M.hat <- matrix(0, n, K)
  M.hat[cbind(1:n, Z.hat)] <- 1
  b.hat <- crossprod(X, M.hat) 
  
  # initialize EM algorithm 
  count.h <- tabulate(Z.hat)
  pi.h <- count.h/n
  
  thet.h <- matrix(0, K, K)
  for(k1 in 1:K) { 
    for(k2 in 1:K) { 
      thet.h[k1, k2] <- mean(X[Z.hat == k1, Z.hat == k2, drop=F])  
    }  
  }
  
  lam.h <- n*t(thet.h*pi.h)
  eta.h <- lam.h/rowSums2(lam.h)
  eta.h[eta.h == 0] <- 1e-16
  
  converged <- F
  iter <- 1
  while((iter <= max.iter) && !converged) { 
    # Compute objective function 
    log.phi.h <- tcrossprod(b.hat, log(eta.h))
    log.phi.pi.h <- t(log.phi.h) + log(pi.h) 
    cpl <- sum(colLogSumExps(log.phi.pi.h))
    
    # Check tolerance
    if(iter > 1) { 
      diffcpl <- abs((cplOld - cpl)/cplOld)
      converged <- (diffcpl < tol)
    }
    
    # E-step 
    log.phi.pi.h.norm <- t(log.phi.pi.h) - colMaxs(log.phi.pi.h)
    gam.h <- exp(log.phi.pi.h.norm)/rowSums2(exp(log.phi.pi.h.norm))
    
    # M-step 
    pi.h <- colMeans2(gam.h)
    eta.h <- t(gam.h)%*%b.hat
    eta.h <- eta.h/rowSums2(eta.h)
    eta.h[eta.h == 0] <- 1e-16
    
    cplOld <- cpl
    iter <- iter + 1
  }
  
  # Estimates from EM
  Z.hat <- max.col(gam.h)
  log.phi.h <- tcrossprod(b.hat, log(eta.h))
  log.phi.pi.h <- t(log.phi.h) + log(pi.h) 
  cpl.est <- sum(colLogSumExps(log.phi.pi.h))
  
  # Deal with any zero degree nodes
  if(sum(zd) > 0) { 
    log.phi.h.final <- matrix(NA, orig.n, K)
    log.phi.h.final[!zd, ] <- log.phi.h
    log.phi.h.final[zd, ] <- 0
    
    log.phi.pi.h <- t(log.phi.h) + log(pi.h) 
    log.phi.pi.h.norm <- t(log.phi.pi.h) - colMaxs(log.phi.pi.h)
    
    gam.h.final <- matrix(NA, orig.n, K) 
    gam.h.final[!zd, ] <- exp(log.phi.pi.h.norm)/rowSums2(exp(log.phi.pi.h.norm))
    gam.h.final[zd, ] <- pi.h
    
    Z.hat.final <- rep(NA, orig.n)
    Z.hat.final[!zd] <- Z.hat
    # Randomly assign hard clusters for the zero degree nodes 
    Z.hat.final[zd] <- sample(1:K, sum(zd), replace=T, prob=pi.h)
    
    Z.hat <- Z.hat.final
    log.phi.h <- log.phi.h.final
    gam.h <- gam.h.final
    cpl.est <-  sum(colLogSumExps(t(log.phi.h.final) + log(pi.h) ))
  }
  
  return(list(logphi=log.phi.h, pi=pi.h, eta=eta.h, class=Z.hat, 
              responsibilities=gam.h, ploglik=cpl.est, converged=converged))
}
  